### PR TITLE
Add Richaudience bidder

### PR DESCRIPTION
**Pull request by CTMBNara**

### CODE REVIEW

File: src/main/java/org/prebid/server/bidder/richaudience/RichaudienceBidder.java [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+    private final JacksonMapper mapper;
+
+    public RichaudienceBidder(String endpointUrl, JacksonMapper mapper) {
+        this.endpointUrl = HttpUtil.validateUrl(Objects.requireNonNull(endpointUrl));
+        this.mapper = Objects.requireNonNull(mapper);
+    }
+
+    @Override
+    public Result<List<HttpRequest<BidRequest>>> makeHttpRequests(BidRequest request) {
+        if (!isDeviseIpPresent(request.getDevice())) {
```
[the remark this answers is not in the document]
Comment (snahornyi) [reply src/main/java/org/prebid/server/bidder/richaudience/RichaudienceBidder.java:54]: This can be extracted to validate request method

File: src/main/java/org/prebid/server/bidder/richaudience/RichaudienceBidder.java [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+        }
+
+        final List<Imp> modifiedImps = new ArrayList<>();
+        final List<BidderError> errors = new ArrayList<>();
+        boolean isTest = false;
+
+        for (Imp imp : request.getImp()) {
+            if (!isBannerCorrect(imp.getBanner())) {
+                final String errorMessage = String.format("Banner W/H/Format is required. ImpId: %s", imp.getId());
+                errors.add(BidderError.badInput(errorMessage));
```
Comment (snahornyi) [inline src/main/java/org/prebid/server/bidder/richaudience/RichaudienceBidder.java:72]: From Go. They fail on first error gotten

File: src/main/java/org/prebid/server/bidder/richaudience/RichaudienceBidder.java [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+        return Result.of(Collections.singletonList(createHttpRequest(modifiedRequest)), errors);
+    }
+
+    private static boolean isDeviseIpPresent(Device device) {
+        return device != null
+                && (StringUtils.isNotBlank(device.getIp())
+                || StringUtils.isNotBlank(device.getIpv6()));
+    }
+
+    private static boolean isBannerCorrect(Banner banner) {
```
Comment (snahornyi) [inline src/main/java/org/prebid/server/bidder/richaudience/RichaudienceBidder.java:100]: isBannerSizesPresent. But actually don't create methods for if conditions.

File: src/test/resources/org/prebid/server/it/test-application.properties [hunk 4]
```diff
@@ -1,5 +1,3 @@
-adapters.generic.enabled=true
```
Comment (snahornyi) [inline src/test/resources/org/prebid/server/it/test-application.properties:1]: Changes like this should be in another PRs. Pls add only bidder specific change and remove others

File: src/main/java/org/prebid/server/bidder/richaudience/RichaudienceBidder.java [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+
+            for (Imp imp : request.getImp()) {
+                validateImp(imp);
+
+                final ExtImpRichaudience richaudienceImp = parseImpExt(imp);
+                modifiedImps.add(modifyImp(imp, richaudienceImp, isSecure));
+
+                if (!isTest && BooleanUtils.isTrue(richaudienceImp.getTest())) {
+                    isTest = true;
+                }
```
Comment (And1sS) [inline src/main/java/org/prebid/server/bidder/richaudience/RichaudienceBidder.java:72]: Can be simplified to `isTest |= BooleanUtils.isTrue(richaudienceImp.getTest());`

**Review by And1sS — CHANGES_REQUESTED** [review]

File: src/main/java/org/prebid/server/bidder/richaudience/RichaudienceBidder.java [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+        } catch (PreBidException e) {
+            return Result.withError(BidderError.badInput(e.getMessage()));
+        }
+    }
+
+    private static void validateRequest(BidRequest bidRequest) throws PreBidException {
+        final Device device = bidRequest.getDevice();
+        final boolean isDeviceValid = device != null
+                && (StringUtils.isNotBlank(device.getIp())
+                || StringUtils.isNotBlank(device.getIpv6()));
```
[the remark this answers is not in the document]
Comment (CTMBNara) [reply src/main/java/org/prebid/server/bidder/richaudience/RichaudienceBidder.java:92]:
```
StringUtils.isAnyBlank(...)
```

File: src/main/java/org/prebid/server/bidder/richaudience/RichaudienceBidder.java [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+            return null;
+        }
+    }
+
+    private static void validateImp(Imp imp) throws PreBidException {
+        final Banner banner = imp.getBanner();
+        final boolean isBannerSizesPresent = banner != null
+                && (banner.getW() != null || banner.getH() != null
+                || CollectionUtils.isNotEmpty(banner.getFormat()));
+        if (!isBannerSizesPresent) {
```
[the remark this answers is not in the document]
Comment (CTMBNara) [reply src/main/java/org/prebid/server/bidder/richaudience/RichaudienceBidder.java:111]: Nope :p

File: src/main/java/org/prebid/server/bidder/richaudience/RichaudienceBidder.java (hunk 5 above)
Comment (CTMBNara) [reply src/main/java/org/prebid/server/bidder/richaudience/RichaudienceBidder.java:72]: Bitwise operators will not be optimized

Comment (And1sS) [reply src/main/java/org/prebid/server/bidder/richaudience/RichaudienceBidder.java:72]: Not a huge performance hit, but huge readability bump.

**Review by And1sS — APPROVED** [review]

**Review by And1sS — APPROVED** [review]

File: src/main/java/org/prebid/server/bidder/richaudience/RichaudienceBidder.java [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+                modifiedImps.add(modifyImp(imp, extImpRichaudience, isSecure));
+
+                isTest |= BooleanUtils.isTrue(extImpRichaudience.getTest());
+            }
+        } catch (PreBidException e) {
+            return Result.withError(BidderError.badInput(e.getMessage()));
+        }
+
+        final BidRequest modifiedRequest = modifyBidRequest(request, url, modifiedImps, isTest);
+        return Result.withValues(Collections.singletonList(createHttpRequest(modifiedRequest)));
```
Comment (And1sS) [inline src/main/java/org/prebid/server/bidder/richaudience/RichaudienceBidder.java:77]: Can be simplified with `.withValue()`

**Review by And1sS — CHANGES_REQUESTED** [review]

**Review by And1sS — APPROVED** [review]

File: src/main/java/org/prebid/server/bidder/richaudience/RichaudienceBidder.java [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+
+        try {
+            validateRequest(request);
+            for (Imp imp : request.getImp()) {
+                validateImp(imp);
+
+                final ExtImpRichaudience extImpRichaudience = parseImpExt(imp);
+                modifiedImps.add(modifyImp(imp, extImpRichaudience, isSecure));
+
+                isTest |= BooleanUtils.isTrue(extImpRichaudience.getTest());
```
Comment (snahornyi) [inline src/main/java/org/prebid/server/bidder/richaudience/RichaudienceBidder.java:70]: Pls reWrite `|=`. We should not use code that can confuse anyone

File: src/main/java/org/prebid/server/bidder/richaudience/RichaudienceBidder.java [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+            return Result.withError(BidderError.badInput(e.getMessage()));
+        }
+
+        final BidRequest modifiedRequest = modifyBidRequest(request, url, modifiedImps, isTest);
+        return Result.withValue(createHttpRequest(modifiedRequest));
+    }
+
+    private static void validateRequest(BidRequest bidRequest) throws PreBidException {
+        final Device device = bidRequest.getDevice();
+        final boolean isDeviceValid = device != null && !StringUtils.isAllBlank(device.getIp(), device.getIpv6());
```
Comment (snahornyi) [inline src/main/java/org/prebid/server/bidder/richaudience/RichaudienceBidder.java:82]: Let's get rid of this variable and put it directly to if statement

File: src/main/java/org/prebid/server/bidder/richaudience/RichaudienceBidder.java [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+        } catch (MalformedURLException e) {
+            return null;
+        }
+    }
+
+    private static void validateImp(Imp imp) throws PreBidException {
+        final Banner banner = imp.getBanner();
+        final boolean isBannerSizesPresent = banner != null
+                && (ObjectUtils.anyNotNull(banner.getW(), banner.getH())
+                || CollectionUtils.isNotEmpty(banner.getFormat()));
```
Comment (snahornyi) [inline src/main/java/org/prebid/server/bidder/richaudience/RichaudienceBidder.java:100]: For this let's create method `isBannerSizesPresent` and invoke it directly in if statement

File: src/main/java/org/prebid/server/bidder/richaudience/RichaudienceBidder.java [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+        if (!isBannerSizesPresent) {
+            final String errorMessage = String.format("Banner W/H/Format is required. ImpId: %s", imp.getId());
+            throw new PreBidException(errorMessage);
```
Comment (snahornyi) [inline src/main/java/org/prebid/server/bidder/richaudience/RichaudienceBidder.java:103]: Pls get rid of errorMessage and put it's value into exception constructor

File: src/test/java/org/prebid/server/bidder/richaudience/RichaudienceBidderTest.java [hunk 13]
```diff
@@ -0,0 +1,280 @@
...
+        return impCustomizer.apply(builder).build();
+    }
+
+    private Imp givenImp(UnaryOperator<Imp.ImpBuilder> impCustomizer) {
+        return givenImp(identity(), impCustomizer);
+    }
+
+    private HttpCall<BidRequest> givenHttpCall(String body) {
+        final HttpResponse response = HttpResponse.of(200, null, body);
+        return HttpCall.success(null, response, null);
```
Comment (snahornyi) [inline src/test/java/org/prebid/server/bidder/richaudience/RichaudienceBidderTest.java:278]: Nitpick. Pls use \n before return statement(except situations where return statement is only the one line in method). Also pls check similar

**Review by snahornyi — CHANGES_REQUESTED** [review]

File: src/main/java/org/prebid/server/bidder/richaudience/RichaudienceBidder.java (hunk 9 above)
Comment (CTMBNara) [reply src/main/java/org/prebid/server/bidder/richaudience/RichaudienceBidder.java:70]: How does compound assignment operator can confuse someone?

**Review by And1sS — DISMISSED** [review]

File: src/main/java/org/prebid/server/bidder/richaudience/RichaudienceBidder.java (hunk 9 above)
Comment (snahornyi) [reply src/main/java/org/prebid/server/bidder/richaudience/RichaudienceBidder.java:70]: It can, project code is not a place where you show how "great programmer you are" and how you can do it. It's a place in which other programmers will come and read code. So pls, do it to be easy to read for every one.

**Review by snahornyi — CHANGES_REQUESTED** [review]

File: src/main/java/org/prebid/server/bidder/richaudience/RichaudienceBidder.java (hunk 9 above)
Comment (snahornyi) [reply src/main/java/org/prebid/server/bidder/richaudience/RichaudienceBidder.java]: For your case this one is even not needed just `=` will be enough

File: src/main/java/org/prebid/server/bidder/richaudience/RichaudienceBidder.java [hunk 14]
```diff
@@ -0,0 +1,177 @@
...
+            validateRequest(request);
+            for (Imp imp : request.getImp()) {
+                validateImp(imp);
+
+                final ExtImpRichaudience extImpRichaudience = parseImpExt(imp);
+                modifiedImps.add(modifyImp(imp, extImpRichaudience, isSecure));
+
+                if (!isTest && BooleanUtils.isTrue(extImpRichaudience.getTest())) {
+                    isTest = true;
+                }
```
Comment (And1sS) [inline src/main/java/org/prebid/server/bidder/richaudience/RichaudienceBidder.java:72]: looks like it can be simplified to `isTest |= BooleanUtils.isTrue(extImpRichaudience.getTest());`

**Review by And1sS — CHANGES_REQUESTED** [review]
